build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1185,6 +1371,140 @@ message ExecuteRequest {
+  // outputs from the action, however in this case the server MUST NOT return
+  // any incremental messages in this field.
+  repeated ConsoleOutput streamed_console_outputs = 3;
+
+  // A directory path for which the worker does not have metadata. this
+  // indicates a directory that the `Action` was not predicted to use
+  // from the client filesystem. The client is requested to enumerate
+  // directory metadata and respond in
+  // [StreamedDynamicExecutionRequest.dynamic_directory_metadata_response][build.bazel.remote.execution.StreamedDynamicExecutionRequest].
+  // The client MAY include individual file digests but this is not required,
+  // the worker will request individual file hashes as needed using
+  // `dynamic_file_hash_path`.
+  string dynamic_directory_content_path = 4;
+
+  // A file path for which the worker does not have a hash. This indicates
+  // a file that the `Action` was not predicted to use from the client
+  // filesystem. The client is requested to hash the file and respond in
+  // [StreamedDynamicExecutionRequest.dynamic_file_hash_response][build.bazel.remote.execution.StreamedDynamicExecutionRequest].
+  string dynamic_file_hash_path = 5;
+
+  // A file path and expected content hash that the worker does not have in
+  // the CAS. This indicates file content that the `Action` was not predicted
+  // to use from the client, or content that was subsequently removed from
+  // the CAS. The client is requested to upload the file content into the
+  // CAS and respond in 
+  // [StreamedDynamicExecutionRequest.dynamic_file_content_response][build.bazel.remote.execution.StreamedDynamicExecutionRequest].
+  DynamicFileContentRequest dynamic_file_content_request = 6;
+}
+
+// A streaming request message for
+// [Execution.ExecuteWithDynamicRetrieval][build.bazel.remote.execution.v2.Execution.ExecuteWithDynamicRetrieval].
+message StreamedDynamicExecutionRequest {
+  // Sent in response to a dynamic directory metadata request.
+  message DynamicDirectoryMetadataResponse {
+    // The directory path, copied from the directory metadata request.
+    string path = 1;
+
+    // Metadata for the directory. [DirectoryNode][build.bazel.remote.execution.v2.Execution.DirectoryNode]
+    // instances in this field MAY omit digest reference information, as
+    // the directory name is the required information needed to feed into
+    // the filesystem, and the contents of the subdirectory contained in the
+    // `DirectoryNode` are not required to be in the CAS. If the
+    // `DirectoryNode` digest is sent to the worker, the referenced `Directory`
+    // information MUST have been previously uploaded to the CAS.
+    Directory directory_metadata = 2;
+  }
+
+  // Sent in response to a dynamic file hash request.
+  message DynamicFileHashResponse {
+    // The file path, copied from the file hash request.
+    string path = 1;
+    
+    // The file digest and associated metadata.
+    FileNode file = 2;
```

Comment:
Does it support the case if the file doesn't exist at the client? For example, the VFS pre-rendering based on historical info may request a file that doesn't exist at client.

Reply:
Fixed

---
In reply to: [459012835](https://github.com/erikma/remote-apis/pull/1#discussion_r459012835) [](ancestors = 459012835)